### PR TITLE
feat: add Address trait to standardize content_addr()

### LINKED ISSUE #163: Make slot access consistent in ASM for dec vars, pub dec vars and state
Currently the ASM to access the dec var and state slots are all accessing the same structure of data `Vec<Vec<Word>>`. However these ops are inconsistent.
Note that I will refer to the inner `Vec<Word>` as `Value` so `slots` are `Vec<Value>`. Each `slot` is a `Value`.

Pub dec vars (transient data) is a bit different. The underlying data structure is `HashMap<u16, HashMap<Vec<Word>, Vec<Word>>>` or `HashMap<SolutionDataIndex, HashMap<Key, Value>>`.
There might still be some things about this asm that we can make more consistent.

# Current situation
The current [ASM](https://github.com/essential-contributions/essential-base/blob/f07cffaf65924a99e90e2f67f38f0da8bfe8fc68/crates/asm-spec/asm.yml#L208) is:

*All ops panic if the indices are out of range unless specified otherwise.*

## Decision Vars

```yaml
# Returns the first word from the slot at `slot_ix`. `slots[slot_ix][0]`.
DecisionVar:
  stack_in: [slot_ix]
  stack_out: [word]

# Returns the a single word at `value_ix` from the slot at `slot_ix`. `slots[slot_ix][value_ix]`.
DecisionVarAt:
  stack_in: [slot_ix, value_ix]
  stack_out: [word]

# Returns the a len words starting at at `value_ix` from the slot at `slot_ix`. `slots[slot_ix][value_ix..(value_ix + len)]`.
DecisionVarRange:
  stack_in: [slot_ix, value_ix, len]
  stack_out:
    elem: word
    len: len

# Returns the length of a slot at `slot_ix`. `slots[slot_ix].len()`.
DecisionVarLen:
  stack_in: [slot_ix]
  stack_out: [len]
```

## State Slots

```yaml
# Returns the all the words from the slot at `slot_ix`. `slots[slot_ix]`.
# Note if the slot is empty this is a noop.
State:
  stack_in: [slot_ix, delta]
  stack_out:
    elem: word
    len: slots[slot_ix].len()

# Returns the all the words from a range of slots at `slot_ix`. `slots[slot_ix..(slot_id + len)]`.
StateRange:
  stack_in: [slot_ix, len, delta]
  stack_out:
    elem: word
    len: len

# Returns the length of a slot at `slot_ix`. `slots[slot_ix].len()`.
StateLen:
  stack_in: [slot_ix, delta]
  stack_out: [len]

# Returns the lengths of slots from a range of slots at `slot_ix`. `slots[slot_ix..(slot_id + len)].iter().map(|s| s.len())`.
StateLenRange:
  stack_in: [slot_ix, len, delta]
  stack_out:
    elem: slot_len
    len: len
```

## Pub Dec Vars (transient)

```yaml
# Returns the all the words at the `key` for the `pathway_ix`. `map[pathway_ix][key]`.
# Note if the value is empty this is a noop.
Transient:
  stack_in: [key_0, ...key_N, key_len, pathway]
  stack_out:
    elem: word
    len: map[pathway_ix][key].len()

# Returns the length of the value at the `key` for the `pathway_ix`. `map[pathway_ix][key].len()`.
TransientLen:
  stack_in: [key_0, ...key_N, key_len, pathway]
  stack_out: [len]

# Returns the number of keys in the map for the current predicate. `map[this_pathway_ix].len()`.
ThisTransientLen:
  stack_out: [len]

# Returns `true` if the map at the current predicate contains the key. `map[this_pathway_ix].contains(key)`.
ThisTransientContains:
  stack_in: [key_0, ...key_N, key_len]
  stack_out: [bool]
```

# Make ops consistent 

## Read full values

First question is should we have ops that read the entire value (even if it's empty)?.
This is useful for things like:
```asm
push 0
push 0
state
push 2
push 0
jump_if
push 42
eq
```
But the same could be achieved by just jumping over the state read as well.

The two ops that currently do this are `Transient` and `State`. 
If we choose to keep this then I think we should have an equivalent for decision vars.
If we choose to remove it then `State` and `Transient` would need to also take an expected length of the value and panic if it doesn't match.
It's not clear to me which is better, I think the Pint team might have a better idea.

## Read first word of value
I think the `DecisionVar` is a bit of a weird specialized op. Is it really that often that you expect a value to be a single word. If so then maybe it's worth having this but noting it is an optimization. I think we should at least rename it to `DecisionVarFirst` or something. If we keep it we should add the same to state and pub dec vars.

## Pub dec var vs transient
We should just choose one of these names and stick to it. Make all the ops match. Pub dec vars has grown on me. The only thing that feels a bit weird about it is that dec vars are always index by a single word where as pub dec vars are indexed by a key. It might be a bit weird equating the two like this.

## Proposal
I think we should have the following core ops:

### Decision Vars

```yaml
# Returns the a len words starting at at `value_ix` from the slot at `slot_ix`. `slots[slot_ix][value_ix..(value_ix + len)]`.
DecisionVar:
  stack_in: [slot_ix, value_ix, len]
  stack_out:
    elem: word
    len: len

# Returns the length of a slot at `slot_ix`. `slots[slot_ix].len()`.
DecisionVarLen:
  stack_in: [slot_ix]
  stack_out: [len]
```

### State Slots

```yaml
# Returns the all the words from a range of slots at `slot_ix`. `slots[slot_ix..(slot_id + len)]`.
State:
  stack_in: [slot_ix, value_ix, len, delta]
  stack_out:
    elem: word
    len: len

# Returns the length of a slot at `slot_ix`. `slots[slot_ix].len()`.
StateLen:
  stack_in: [slot_ix, delta]
  stack_out: [len]
```

### Pub Dec Vars (transient)

```yaml
# Returns the all the words at the `key` for the `pathway_ix`. `map[pathway_ix][key]`.
PubVar:
  stack_in: [pathway, key_0, ...key_N, key_len, value_ix, len]
  stack_out:
    elem: word
    len: len

# Returns the length of the value at the `key` for the `pathway_ix`. `map[pathway_ix][key].len()`.
PubVarLen:
  stack_in: [key_0, ...key_N, key_len, pathway]
  stack_out: [len]
```

Then leave room at the end of the `Access` byte range to add in more optimized ops like `DecVarFirst` or `PubVarFullValue` as the need actually arises. Considering `Access` is already very full we could reserve the next byte for additional `Access` ops.

`ThisTransientLen` and `ThisTransientContains` should follow the form of `MutKeys` and become:
```yaml
# Returns the set of keys for the pub var map at a `pathway_ix`. `map[pathway_ix].keys().collect::<HashSet<_>>()`.
PubVarKeys:
  stack_in: [pathway_ix]
  stack_out: [key_0, key_0_len, ...key_N, key_N_len, total_length]
``` 
That way they can be restricted in the same way as mut keys as follows:
```asm
push 42
push 1
push 1
this_pathway
pub_var_keys
eq_set
```

### PR DESCRIPTION
**Pull request by freesig**
closes https://github.com/essential-contributions/essential-node/issues/39

Adds the `Address` trait so that the `essential_hash::content_addr()` function will always produce the correct hash.

### CODE REVIEW

File: crates/hash/src/lib.rs [hunk 1]
```diff
@@ -8,8 +8,19 @@ use essential_types::{convert::bytes_from_word, ContentAddress, Hash, Word};
...
 pub mod contract_addr;
 
+/// Standardized trait for creating content addresses for
+/// types using the correct constructors.
+pub trait Address: Serialize {
+    /// Produce the content address for self.
+    fn content_address(&self) -> ContentAddress {
+        ContentAddress(hash(&self))
+    }
+}
```
Comment (mitchmindtree) [inline crates/hash/src/lib.rs:22]: I'd probably remove the `Serialize` bound and the default impl, then add the implementations manually for `Solution` and `Predicate`, as serialization is not strictly related.
```suggestion
/// Standardized trait for creating content addresses for
/// types using the correct constructors.
pub trait Address {
    /// Produce the content address for self.
    fn content_address(&self) -> ContentAddress;
}
```

File: crates/hash/src/block_addr.rs [hunk 2]
```diff
@@ -0,0 +1,53 @@
+//! A small collection of helper functions to assist in the calculation of an
+//! block's content address.
```
Comment (mitchmindtree) [inline crates/hash/src/block_addr.rs:2]:
```suggestion
//! A small collection of helper functions to assist in the calculation of a
//! block's content address.
```

File: crates/hash/src/block_addr.rs [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+//!
+//! Note that it is possible this hash will change in the future
+//! if merkle proofs or a header type is introduced.
+
+use essential_types::{Block, ContentAddress};
+
+/// Shorthand for the common case of producing an block's content address
```
Comment (mitchmindtree) [inline crates/hash/src/block_addr.rs:9]:
```suggestion
/// Shorthand for the common case of producing a block's content address
```

File: crates/check/src/solution.rs [hunk 4]
```diff
@@ -213,7 +213,7 @@ impl<E: fmt::Display> fmt::Display for PredicateErrors<E> {
 /// its associated predicates.
 ///
 /// This includes solution data and state mutations.
-#[cfg_attr(feature = "tracing", tracing::instrument(skip_all, fields(solution = %content_addr(&solution.data)), err))]
+#[cfg_attr(feature = "tracing", tracing::instrument(skip_all, fields(solution = %content_addr(solution)), err))]
```
Comment (mitchmindtree) [inline crates/check/src/solution.rs:216]: wow new trait already doing work

**Review by mitchmindtree — APPROVED** [review]
Flip yess, sooo much nicer not having to do `contract_addr::from_contract` all the time for contracts.

I've left one minor tweak I'd make to the trait, otherwise looking great.

File: crates/hash/src/lib.rs (hunk 1 above)
Comment (freesig) [reply crates/hash/src/lib.rs:22]: Isn't `Serialize` always required by `essential_hash::hash()`? Are you saying someone might want to implement this using some other underlying hash scheme that has some custom serialization (i.e. not serde)? If so do we really want to be that flexible?

**Pull request merged**